### ISSUE TITLE
Environment setup error and LoRA finetuning baselines

### ISSUE THREAD

**Issue by bing0037** [InternLM/xtuner#259]
Hi, thanks for your work. Recently I wanna to reproduce your results (SFT baseline and lora performance).

# Setup env:
I followed the [README](https://github.com/InternLM/xtuner/blob/main/README.md) to setup the env. The env is setup with no error.

# Test 1: use xtuner train:
```
xtuner train llama_7b_qlora_alpaca_e3 --deepspeed deepspeed_zero2
```
## Error: How to reproduce the test?
![微信图片_20231208094708](https://github.com/InternLM/xtuner/assets/11786011/2c29d953-d33e-4463-b258-93984eb82057)

# Test 2: use examples:
I tried the examples by following the [README](https://github.com/InternLM/xtuner/blob/main/examples/README.md).
```
python train_lora_hf.py
```

## Question 1: what is the lora baseline? how to choose the hyperparameters (e.g. learning_rate, etc.) ? 
On alpaca dataset, using default code (learning_rate=2e-5, num_train_epochs=3), the final training loss is 1.03; using custom configuration (learning_rate=2e-5), the final training loss is 0.77. What is the best configuration for this dataset? Also what to choose for other datasets (e.g. sql, arxiv, code, medical, etc.)?

## Question 2: what is the original full-parameters (without lora) finetuning baseline?
By replacing `build_lora_model()` by `build_model()` in [train_lora_hf.py](https://github.com/InternLM/xtuner/blob/main/examples/train_lora_hf.py), I was trying to reproduce the full-paramaters finetuning (without lora). **The training loss is always 0**. There should be something wrong. How could I reproduce the full-parameters finetuning baseline?
![微信图片_20231208094842](https://github.com/InternLM/xtuner/assets/11786011/0b4545e4-0f2a-4b35-9720-9f3205826231)

## Question 3: how to enable deepspeed ZeRO while using examples/train_lora_hf.py?
I wanna to reproduce the full-parameters finetuning baseline on different LLMs. With ZeRO, I will be able to train larger LLM with fewer GPUs. Could you give some suggestions?

Many thanks!

**Comment by LZHgrla** [comment]
@bing0037  Hi, thanks!

**Test 1**
I cannot repro this error, with `llama_7b_qlora_alpaca_e3`. Did you make any modifications to the config? If so, please post your config!

**Test 2**
*It is noted that we do not produce models. Our framework is intended to provide a repository with efficient fine-tuning capability.*

**Q1**
We recommend a learning rate of 2e-4 for QLoRA and LoRA, and 2e-5 for full fine-tuning. Other parameters, such as batch size, epoch, and weight decay, can be tuned according to your needs.

You can use one GPU training with https://github.com/LZHgrla/xtuner/blob/main/xtuner/configs/llama/llama_7b/llama_7b_qlora_oasst1_e3.py. The model trained with this is comparable to the QLoRA official models when evaluated on MMLU. Since evaluating only based on MMLU is one-sided, we did not add it to all the configs. If you consider enabling it, you can refer to PR https://github.com/InternLM/xtuner/pull/208 and [config](https://github.com/LZHgrla/xtuner/blob/main/xtuner/configs/internlm/internlm_7b/internlm_7b_qlora_oasst1_mmlu_e3.py).

**Q2**
Maybe you should reduce the learning rate for full-parameter fine-tuning.

**Q3**
In examples, we aim to demonstrate the usage of our `xtuner.apis`. It is only used to provide models and datasets, while training relies on `transformers.Trainer`. So, you can refer the docs of `transformers.Trainer` to obtain more usage. [DOCS](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.deepspeed)

If you want to use ZeRO, we recommend using xtuner's training pipeline, as you did in Test 1! (So, the first thing is to work together to resolve your error bug!)

**Comment by bing0037** [comment]
@LZHgrla Thanks for your reply.

I reinstall the env from source and currently the `xtuner train` running works well.

I run the alpaca finetuning test using [llama_7b_qlora_alpaca_e3.py](https://github.com/InternLM/xtuner/blob/main/xtuner/configs/llama/llama_7b/llama_7b_qlora_alpaca_e3.py) with default lr=2e-5. The final training loss is 1.02. Then I tried lr=2e-5, the final training loss is 1.09.

(1) Could you provide the expected training loss on alpaca dataset (lora & qlora & full-parameter)? 

(2) Also, could you provide the config to perform full-parameter finetuning to reproduce the full-parameter baseline (hyperparameter selection: lr / epochs / deepspeed zero / ...)?

(3) More baseline results are appreciated (alpaca / arxiv / code/ lawyer /... with full-parameter/lora/qlora).

Thanks!

**Comment by LZHgrla** [comment]
@bing0037 
The configs we provide are all for single-gpu fine-tuning (except those specifically marked `***_gpu8.py`). If there is no error with your environment, then the result obtained by directly using `xtuner train $CONFIG` can be considered as the baseline.
We have some models released on [Huggingface](https://huggingface.co/xtuner) and [Modelscope](https://www.modelscope.cn/organization/xtuner), but there are no records specifically for loss. The reason is that we are not building baseline models, but providing a repository with capability of efficient fine-tuning LLM.

**Comment by bing0037** [comment]
@LZHgrla Thanks a lot for your reply. 

# test results:
I tested the full-parameter and qlora finetuning on alpaca dataset. 

(1) For qlora, I use the original [llama_7b_qlora_alpaca_e3.py](https://github.com/InternLM/xtuner/blob/main/xtuner/configs/llama/llama_7b/llama_7b_qlora_alpaca_e3.py). 
The final training loss is 1.02.

(2) For the full-parameter finetuning, I modify the config by removing `lora` and `quantization_config` as shown below:
The final training loss is 0.7.

![微信图片_20231211191346](https://github.com/InternLM/xtuner/assets/11786011/c349bc4a-938b-4521-9b87-8604bbeb16ed)

# Question:
Should we suppose to have a comparable result with full-parameter finetuning? Could you provide some information or suggestion on how to get a comparable result using lora or qlora?

**Comment by LZHgrla** [comment]
@bing0037 
If the training loss does not show any outliers, it can be considered as the baseline for this model, data, and corresponding configuration.

The results you've shown seem to be fine, and it is quite normal that the loss when using LoRA or QLoRA is indeed higher than that of full fine-tuning.
